### PR TITLE
BOPs KPI computation

### PR DESCRIPTION
**Pull request by ofirgo**
This is the second PR in the set of PRs which add BOPs KPI to mixed-precision. Need to be merged AFTER the first PR.

### CODE REVIEW

File: model_compression_toolkit/core/common/mixed_precision/kpi_tools/kpi_methods.py [hunk 1]
```diff
@@ -133,6 +159,156 @@ def total_weights_activation_kpi(mp_cfg: List[int], graph: Graph, fw_info: Frame
...
+    if not set_constraints:
+        return _bops_kpi(mp_cfg,
+                         graph,
+                         fw_info,
+                         fw_impl)
+
+    virtual_bops_nodes = [n for n in graph.get_topo_sorted_nodes() if isinstance(n, VirtualActivationWeightsNode)]
+
+    mp_nodes = graph.get_configurable_sorted_nodes_names()
+    bops = [n.get_bops_count(fw_impl, fw_info, candidate_idx=_get_node_cfg_idx(n, mp_cfg, mp_nodes)) for n in virtual_bops_nodes]
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/mixed_precision/kpi_tools/kpi_methods.py:193]: Shouldn't we compute BOPs for the interest points and not configurable nodes?

File: model_compression_toolkit/core/common/graph/virtual_activation_weights_node.py [hunk 2]
```diff
@@ -183,3 +183,21 @@ def __init__(self,
...
+            candidate_idx: The index of the node's quantization candidate configuration.
+
+        Returns: The BOPS count of the composed node.
+
+        """
+        node_mac = fw_impl.get_node_mac_operations(self.original_weights_node, fw_info)
+        node_bops = self.candidates_quantization_cfg[candidate_idx].weights_quantization_cfg.weights_n_bits * \
+                    self.candidates_quantization_cfg[candidate_idx].activation_quantization_cfg.activation_n_bits * \
+                    node_mac
+        return node_bops
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/graph/virtual_activation_weights_node.py:203]: Why not checking if quantization is enabled (if not use FLOAT_BITWIDTH)?

**Review by reuvenperetz — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/core/keras/keras_implementation.py [hunk 3]
```diff
@@ -484,3 +484,37 @@ def is_node_compatible_for_mp_metric_outputs(self,
...
+            return np.prod([x for x in output_shape if x is not None]) * \
+                   input_shape[input_channel_axis] * \
+                   (kernel_shape[0] * kernel_shape[1])
+        elif node.type is DepthwiseConv2D:
+            # Depth * (W_out * H_out) * C_in * (W_kernel * H_kernel)
+            return node.framework_attr.get(DEPTH_MULTIPLIER) * \
+                   np.prod([x for x in output_shape if x is not None]) / output_shape[output_channel_axis] * \
+                   input_shape[input_channel_axis] * \
+                   (kernel_shape[0] * kernel_shape[1])
+        elif node.type is Dense:
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/core/keras/keras_implementation.py:516]: Isn't it the case for an FC layer? (I'm not sure but isn't the kernel size is [Input_size x Output_size] )

File: model_compression_toolkit/core/common/mixed_precision/kpi_tools/kpi_methods.py (hunk 1 above)
Comment (ofirgo) [reply model_compression_toolkit/core/common/mixed_precision/kpi_tools/kpi_methods.py:193]: No, interest points are only used for distance metric computation.
We need to compute the KPI for each node that contributes to it.

File: model_compression_toolkit/core/common/graph/virtual_activation_weights_node.py (hunk 2 above)
Comment (ofirgo) [reply model_compression_toolkit/core/common/graph/virtual_activation_weights_node.py:203]: Not sure to which part of the code this comment is referring

**Review by reuvenperetz — APPROVED** [review]

**Pull request merged**